### PR TITLE
Feature/mtu discovery 2

### PR DESCRIPTION
**Pull request by jthomas43**
based on and uses terminology from https://datatracker.ietf.org/doc/html/rfc8899

### CODE REVIEW

File: include/udx.h [hunk 1]
```diff
@@ -362,11 +382,14 @@ udx_check_timeouts (udx_t *handle);
 int
 udx_stream_init (udx_t *udx, udx_stream_t *handle, uint32_t local_id, udx_stream_close_cb close_cb);
 
+// todo: remove these?
```
Comment (kasperisager) [inline include/udx.h:385]: Could we leave the getter and remove the setter?

File: src/io_posix.c [hunk 2]
```diff
@@ -15,6 +15,44 @@
...
+  if (rc == -1) {
+    close(s);
+    return -1;
+  }
+
+  close(s);
+  return mtu;
+}
+#endif
+
```
Comment (kasperisager) [inline src/io_posix.c:55]: If this covers all platforms then we can put it in a shared `io.c`.

File: src/udx.c [hunk 3]
```diff
@@ -488,6 +488,35 @@ init_stream_packet (udx_packet_t *pkt, int type, udx_stream_t *stream, const uv_
...
+    return 0;
+  }
+  debug_printf("mtu: probeify seq=%d size=%u wanted=%d padding=%d\n", pkt->seq, pkt->size + header_size, wanted_size, padding_size);
+  static char probe_data[256] = "probeprobeprobe......";
+  pkt->bufs[2] = pkt->bufs[1];
+  pkt->bufs[1].len = padding_size;
+  pkt->bufs[1].base = probe_data;
+  pkt->header[3] = padding_size;
+  pkt->bufs_len = 3;
+  // pkt->size += padding_size;
```
Comment (kasperisager) [inline src/udx.c:509]: Leftover?

File: src/udx.c [hunk 4]
```diff
@@ -488,6 +488,35 @@ init_stream_packet (udx_packet_t *pkt, int type, udx_stream_t *stream, const uv_
...
+  assert(pkt->bufs_len == 2);
+  assert(pkt->header[3] == 0);
+  assert(wanted_size > pkt->size);
+  int header_size = (pkt->dest.ss_family == AF_INET ? UDX_IPV4_HEADER_SIZE : UDX_IPV6_HEADER_SIZE) - 20;
+  int padding_size = wanted_size - (pkt->size + (pkt->dest.ss_family == AF_INET ? UDX_IPV4_HEADER_SIZE : UDX_IPV6_HEADER_SIZE) - 20);
+  if (padding_size > 255) {
+    return 0;
+  }
+  debug_printf("mtu: probeify seq=%d size=%u wanted=%d padding=%d\n", pkt->seq, pkt->size + header_size, wanted_size, padding_size);
+  static char probe_data[256] = "probeprobeprobe......";
```
Comment (kasperisager) [inline src/udx.c:503]: Can we leave this empty?

File: src/udx.c [hunk 5]
```diff
@@ -732,6 +789,9 @@ close_maybe (udx_stream_t *stream, int err) {
     stream->on_close(stream, err);
   }
 
+  // uv_timer_stop(&stream->mtu_probe_timer);
```
Comment (kasperisager) [inline src/udx.c:792]: Leftover?

File: src/udx.c [hunk 6]
```diff
@@ -757,13 +817,24 @@ rack_update_reo_wnd (udx_stream_t *stream) {
   return r < stream->srtt ? r : stream->srtt;
 }
 
+static inline int
```
Comment (kasperisager) [inline src/udx.c:820]:
```suggestion
static inline bool
```

File: src/udx.c [hunk 7]
```diff
@@ -757,13 +817,24 @@ rack_update_reo_wnd (udx_stream_t *stream) {
...
+seq_was_probe (udx_stream_t *s, uint32_t seq) {
+  for (int i = 0; i < s->mtu_probe_count; i++) {
+    if (s->mtu_probe_seq[i] == seq) {
+      return 1;
```
Comment (kasperisager) [inline src/udx.c:824]:
```suggestion
      return true;
```

File: src/udx.c [hunk 8]
```diff
@@ -757,13 +817,24 @@ rack_update_reo_wnd (udx_stream_t *stream) {
...
+    }
+  }
+  return 0;
```
Comment (kasperisager) [inline src/udx.c:827]:
```suggestion
  return false;
```

File: src/udx.c [hunk 9]
```diff
@@ -857,6 +953,34 @@ ack_packet (udx_stream_t *stream, uint32_t seq, int sack) {
     return 0;
   }
 
+  if (stream->mtu_state == UDX_MTU_STATE_SEARCH && stream->mtu_probe_count > 0 && seq == stream->mtu_probe_seq[stream->mtu_probe_count - 1]) {
+    debug_printf("mtu: probe acked seq=%d mtu=%d->%d\n", seq, stream->mtu, stream->mtu_probe_size);
+
+    stream->mtu_probe_count = 0;
+    stream->mtu = stream->mtu_probe_size;
+    // uv_timer_stop(&stream->mtu_probe_timer);
```
Comment (kasperisager) [inline src/udx.c:961]: Leftover?

File: src/udx.c [hunk 10]
```diff
@@ -857,6 +953,34 @@ ack_packet (udx_stream_t *stream, uint32_t seq, int sack) {
...
+    if (stream->mtu_probe_size == stream->mtu_max) {
+      stream->mtu_state = UDX_MTU_STATE_SEARCH_COMPLETE;
+      uv_timer_start(&stream->mtu_raise_timer, mtu_raise_timeout, UDX_MTU_RAISE_TIMEOUT_MS, 0);
+
+    } else {
+      stream->mtu_probe_size += UDX_MTU_STEP;
+      if (stream->mtu_probe_size >= stream->mtu_max) {
+        stream->mtu_probe_size = stream->mtu_max;
+      }
+      // send_mtu_probe_packet(stream);
```
Comment (kasperisager) [inline src/udx.c:972]: Leftover?

File: src/udx.c [hunk 11]
```diff
@@ -857,6 +953,34 @@ ack_packet (udx_stream_t *stream, uint32_t seq, int sack) {
...
+      stream->mtu_probe_wanted = 1;
+    }
+  }
+
+  if (stream->mtu_state == UDX_MTU_STATE_BASE || stream->mtu_state == UDX_MTU_STATE_ERROR) {
+    // if a packet is acked we can start probing upwards
+    stream->mtu_state = UDX_MTU_STATE_SEARCH;
+    // send_mtu_probe_packet(stream); // we can send first probe immediately
```
Comment (kasperisager) [inline src/udx.c:980]: Leftover?

File: include/udx.h [hunk 12]
```diff
@@ -187,6 +199,14 @@ struct udx_stream_s {
   udx_stream_drain_cb on_drain;
   udx_stream_close_cb on_close;
 
+  // mtu. RFC8899 5.1.1 and 5.1.3
+  int mtu_state; // MTU_STATE_*
+  int mtu_probe_wanted;
```
Comment (kasperisager) [inline include/udx.h:204]:
```suggestion
  bool mtu_probe_wanted;
```

File: src/udx.c [hunk 13]
```diff
@@ -662,6 +712,13 @@ fill_window (udx_stream_t *stream) {
 
     stream->pkts_inflight++;
     stream->inflight += pkt->size;
+
+    if (stream->mtu_probe_wanted && mtu_probeify_packet(pkt, stream->mtu_probe_size)) {
+      stream->mtu_probe_seq[stream->mtu_probe_count] = pkt->seq;
+      stream->mtu_probe_count++;
+      stream->mtu_probe_wanted = 0;
```
Comment (kasperisager) [inline src/udx.c:719]:
```suggestion
      stream->mtu_probe_wanted = false;
```

File: src/udx.c [hunk 14]
```diff
@@ -789,20 +860,45 @@ rack_detect_loss (udx_stream_t *stream) {
...
+          if (stream->mtu_probe_count >= UDX_MTU_MAX_PROBES) {
+            stream->mtu_state = UDX_MTU_STATE_SEARCH_COMPLETE;
+            debug_printf("mtu: established, mtu=%d", stream->mtu);
+            if (stream->mtu < stream->mtu_max) {
+              debug_printf(", %d<%d. scheduling mtu_raise_timer", stream->mtu, stream->mtu_max);
+              uv_timer_start(&stream->mtu_raise_timer, mtu_raise_timeout, UDX_MTU_RAISE_TIMEOUT_MS, 0);
+            }
+            debug_printf("\n");
+          } else {
+            stream->mtu_probe_wanted = 1;
```
Comment (kasperisager) [inline src/udx.c:879]:
```suggestion
            stream->mtu_probe_wanted = true;
```

File: src/udx.c [hunk 15]
```diff
@@ -857,6 +953,34 @@ ack_packet (udx_stream_t *stream, uint32_t seq, int sack) {
...
+      stream->mtu_state = UDX_MTU_STATE_SEARCH_COMPLETE;
+      uv_timer_start(&stream->mtu_raise_timer, mtu_raise_timeout, UDX_MTU_RAISE_TIMEOUT_MS, 0);
+
+    } else {
+      stream->mtu_probe_size += UDX_MTU_STEP;
+      if (stream->mtu_probe_size >= stream->mtu_max) {
+        stream->mtu_probe_size = stream->mtu_max;
+      }
+      // send_mtu_probe_packet(stream);
+      stream->mtu_probe_wanted = 1;
```
Comment (kasperisager) [inline src/udx.c:973]:
```suggestion
      stream->mtu_probe_wanted = true;
```

File: src/udx.c [hunk 16]
```diff
@@ -857,6 +953,34 @@ ack_packet (udx_stream_t *stream, uint32_t seq, int sack) {
...
+    }
+  }
+
+  if (stream->mtu_state == UDX_MTU_STATE_BASE || stream->mtu_state == UDX_MTU_STATE_ERROR) {
+    // if a packet is acked we can start probing upwards
+    stream->mtu_state = UDX_MTU_STATE_SEARCH;
+    // send_mtu_probe_packet(stream); // we can send first probe immediately
+    stream->mtu_probe_wanted = 1;
```
Comment (kasperisager) [inline src/udx.c:981]:
```suggestion
    stream->mtu_probe_wanted = true;
```

File: src/udx.c [hunk 17]
```diff
@@ -1544,7 +1670,15 @@ udx_stream_init (udx_t *udx, udx_stream_t *handle, uint32_t local_id, udx_stream
   handle->reordering_seen = false;
   handle->retransmitting = 0;
 
-  handle->mtu = UDX_DEFAULT_MTU;
+  handle->mtu = UDX_MTU_BASE;
+  // handle->mtu_state = UDX_MTU_STATE_DISABLED;
```
Comment (kasperisager) [inline src/udx.c:1674]: Leftover?

File: src/udx.c [hunk 18]
```diff
@@ -1544,7 +1670,15 @@ udx_stream_init (udx_t *udx, udx_stream_t *handle, uint32_t local_id, udx_stream
...
+  handle->mtu_state = UDX_MTU_STATE_BASE;
+  handle->mtu_probe_count = 0;
+  handle->mtu_probe_size = UDX_MTU_BASE; // starts with first ack, counts as a confirmation of base
+  handle->mtu_max = UDX_MTU_MAX;         // revised in connect()
+
+  // uv_timer_init(udx->loop, &handle->mtu_probe_timer);
```
Comment (kasperisager) [inline src/udx.c:1680]: Leftover?

File: test/stream-write-read-perf.c [hunk 19]
```diff
@@ -42,14 +43,15 @@ on_read (udx_stream_t *handle, ssize_t read_len, const uv_buf_t *buf) {
   stats.last_read_ms = uv_hrtime() / 1000000;
 
   if (stats.bytes_read == options.size_bytes) {
-    udx_stream_destroy(handle);
+    // udx_stream_destroy(handle);
```
Comment (kasperisager) [inline test/stream-write-read-perf.c:46]: Leftover?

File: test/stream-write-read-perf.c [hunk 20]
```diff
@@ -42,14 +43,15 @@ on_read (udx_stream_t *handle, ssize_t read_len, const uv_buf_t *buf) {
...
+    printf("read all bytes\n");
+    // __builtin_trap();
```
Comment (kasperisager) [inline test/stream-write-read-perf.c:48]: Leftover?

File: test/stream-write-read-perf.c [hunk 21]
```diff
@@ -112,7 +114,8 @@ main () {
 
   printf("generating data ...\n");
 
-  options.size_bytes = 2 * 1024 * 1024 * 1024L;
+  // options.size_bytes = 2 * 1024 * 1024 * 1024L;
```
Comment (kasperisager) [inline test/stream-write-read-perf.c:117]: Leftover?

**Review by kasperisager — CHANGES_REQUESTED** [review]

File: src/udx.c (hunk 4 above)
Comment (jthomas43) [reply src/udx.c]: done

File: src/udx.c (hunk 5 above)
Comment (jthomas43) [reply src/udx.c]: yes

File: test/stream-write-read-perf.c (hunk 19 above)
Comment (jthomas43) [reply test/stream-write-read-perf.c]: moved the stream destroy (for both streams) to the on_ack event.

File: src/udx.c (hunk 3 above)
Comment (jthomas43) [reply src/udx.c]: yes

File: include/udx.h (hunk 1 above)
Comment (jthomas43) [reply include/udx.h]: done

File: src/udx.c (hunk 6 above)
Comment (jthomas43) [reply src/udx.c]: done

File: src/udx.c (hunk 7 above)
Comment (jthomas43) [reply src/udx.c]: done

**Pull request merged**